File: clinical/src/main/java/com/hartwig/actin/clinical/feed/bodyweight/BodyWeightEntry.java
package com.hartwig.actin.clinical.feed.bodyweight;

import java.time.LocalDate;

import com.hartwig.actin.clinical.feed.FeedEntry;

import org.immutables.value.Value;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

@Value.Immutable
@Value.Style(passAnnotations = { NotNull.class, Nullable.class })
public abstract class BodyWeightEntry implements FeedEntry {

    @NotNull
    @Override
    public abstract String subject();

    public abstract double valueQuantityValue();

    @NotNull
    public abstract String valueQuantityUnit();

    @NotNull
    public abstract LocalDate effectiveDateTime();

}

File: clinical/src/main/kotlin/com/hartwig/actin/clinical/feed/bodyweight/BodyWeightEntry.kt
package com.hartwig.actin.clinical.feed.bodyweight

import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import com.hartwig.actin.clinical.feed.EuropeanDecimalDeserializer
import com.hartwig.actin.clinical.feed.FeedEntry
import com.hartwig.actin.clinical.feed.FeedValidation
import com.hartwig.actin.clinical.feed.FeedValidator
import com.hartwig.actin.clinical.feed.JacksonSerializable
import java.time.LocalDateTime

@JacksonSerializable
data class BodyWeightEntry(
    @JsonProperty("subject")
    override val subject: String,

    @JsonProperty("valueQuantity_value")
    @JsonDeserialize(using = EuropeanDecimalDeserializer::class)
    val valueQuantityValue: Double,

    @JsonProperty("valueQuantity_unit")
    val valueQuantityUnit: String,

    @JsonProperty("effectiveDateTime")
    val effectiveDateTime: LocalDateTime
) : FeedEntry

class BodyWeightEntryValidator : FeedValidator<BodyWeightEntry> {
    override fun validate(feed: BodyWeightEntry): FeedValidation {
        return FeedValidation(feed.valueQuantityValue > 0)
    }
}